ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java

```diff
@@ -0,0 +1,231 @@
+        }
+        indexSegment = nextIndexSegment;
+      } else {
+        break;
+      }
+    }
+    if (indexSegment != null) {
+      String logSegmentName = indexSegment.getLogSegmentName();
+      totalValidDataSize += validDataSize;
+      if (validDataSizeBySegments != null) {
+        validDataSizeBySegments.put(logSegmentName, validDataSize);
+      }
+    }
+    return new Pair<>(referenceTimeInSecs, totalValidDataSize);
+  }
+
+  /**
+   * Finds the last eligible index start offset whose last modified time surpases {@code referenceTimeInSecs}
+   * @param referenceTimeInSecs the reference time against which the last eligible index start offset needs to be found
+   * @return the last eligible index start offset whose last modified time surpases {@code referenceTimeInSecs}
+   */
+  private Offset getLastEligibleStartOffsetForDelete(long referenceTimeInSecs) {
+    if (index.indexes.lastEntry().getValue().getLastModifiedTime() < referenceTimeInSecs) {
+      return index.getCurrentEndOffset();
+    }
+    IndexSegment latestIndexSegment = null;
+    for (IndexSegment indexSegment : index.indexes.values()) {
+      if (indexSegment.getLastModifiedTime() < referenceTimeInSecs) {
+        latestIndexSegment = indexSegment;
+      } else {
+        break;
+      }
+    }
+    return latestIndexSegment.getStartOffset();
+  }
+
+  /**
+   * Get valid data size of an index segment
+   * @param indexSegment the {@link IndexSegment} for which valid data size has to be determined
+   * @param forwardSearchSpan the {@link FileSpan} within which search has to be done to determine validity of data
+   * @param referenceTime time in ms used as reference to check for expiration
+   * @return the valid data size of the given index segment
+   * @throws IOException
+   * @throws StoreException
+   */
+  private long getValidDataSizePerIndexSegment(IndexSegment indexSegment, FileSpan forwardSearchSpan,
+      long referenceTime) throws IOException, StoreException {
+    long validSize = 0;
+    List<MessageInfo> messageInfos = new ArrayList<>();
+    indexSegment.getEntriesSince(null, new FindEntriesCondition(Integer.MAX_VALUE), messageInfos, new AtomicLong(0));
+    for (MessageInfo messageInfo : messageInfos) {
+      IndexValue value = index.findKey(messageInfo.getStoreKey(), forwardSearchSpan);
+      if (messageInfo.getExpirationTimeInMs() == Utils.Infinite_Time && !value.isFlagSet(
+          IndexValue.Flags.Delete_Index)) {
+        // put record w/o any expiration and not deleted
+        validSize += messageInfo.getSize();
+      } else if (!messageInfo.isDeleted() && !(isExpired(messageInfo.getExpirationTimeInMs(), referenceTime)
+          || value.isFlagSet(IndexValue.Flags.Delete_Index))) {
+        // a put record either expired or deleted(in future index segment) within referenceTimeInMs
+        validSize += messageInfo.getSize();
```

Comment:
Why for all three cases `validSize += messageInfo.getSize()`? Especially for the first (non-deleted and non-expired put) and second case (deleted or expired put), it seems the first case should add the total size, and the second case should also add the record size.